### PR TITLE
[Inventory] Decrease quantity of items for given product variant when order is paid

### PR DESCRIPTION
**Pull request by tuka217**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Related tickets | -
| License         | MIT

### CODE REVIEW

**Comment by gabiudrescu** [comment]
omg, I'm soooo following this

File: src/Sylius/Behat/Context/Transform/ProductVariantContext.php [hunk 1]
```diff
@@ -41,6 +41,20 @@ public function __construct(ProductRepositoryInterface $productRepository, Produ
     }
 
     /**
+     * @Transform :variant
+     * @Transform /^in variant "["^]+"$/
+     */
+    public function getProductVariantByName($variantName)
+    {
+        $productVariant = $this->productVariantRepository->findOneBy(['name' => $variantName]);
+        if (null === $productVariant) {
```
Comment (Arminek) [inline src/Sylius/Behat/Context/Transform/ProductVariantContext.php]: Assert::null

File: src/Sylius/Component/Core/Inventory/Updater/OnHandQuantityUpdater.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+namespace Sylius\Component\Core\Inventory\Updater;
+
+use Sylius\Component\Core\Model\OrderItemInterface;
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+class OnHandQuantityUpdater implements OnHandQuantityUpdaterInterface
```
Comment (Arminek) [inline src/Sylius/Component/Core/Inventory/Updater/OnHandQuantityUpdater.php]: ``final``

File: src/Sylius/Component/Core/Inventory/Updater/OnHandQuantityUpdater.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function decrease($orderItems)
+    {
+        foreach ($orderItems as $orderItem) {
+            if($orderItem->getQuantity() < 0) {
```
Comment (Arminek) [inline src/Sylius/Component/Core/Inventory/Updater/OnHandQuantityUpdater.php]: ``Assert::``

File: src/Sylius/Behat/Context/Ui/Shop/CartContext.php [hunk 4]
```diff
@@ -280,15 +280,15 @@ public function iAddMultipleProductsToTheCart(array $products)
...
+     * @When I add :variantName variant of product :product to the cart
+     * @When I have :variantName variant of product :product in the cart
      * @When /^I add "([^"]+)" variant of (this product) to the cart$/
      */
-    public function iAddProductToTheCartSelectingVariant($variant, ProductInterface $product)
+    public function iAddProductToTheCartSelectingVariant($variantName, ProductInterface $product)
     {
         $this->productShowPage->open(['slug' => $product->getSlug()]);
-        $this->productShowPage->addToCartWithVariant($variant);
+        $this->productShowPage->addToCartWithVariant($variantName);
```
Comment (NeverResponse) [inline src/Sylius/Behat/Context/Ui/Shop/CartContext.php]: You could've gone with transforming the variant and using `$variant->getName()` imo

File: src/Sylius/Behat/Resources/config/suites/ui/inventory/inventory.yml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+                - sylius.behat.context.transform.shipping_method
+
+                - sylius.behat.context.setup.admin_security
+                - sylius.behat.context.setup.channel
+                - sylius.behat.context.setup.order
+                - sylius.behat.context.setup.payment
+                - sylius.behat.context.setup.product
+                - sylius.behat.context.setup.shipping
+
+                - sylius.behat.context.transform.shared_storage
```
Comment (NeverResponse) [inline src/Sylius/Behat/Resources/config/suites/ui/inventory/inventory.yml]: Why is this one so far down?

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/state_machine.yml [hunk 6]
```diff
@@ -116,6 +122,7 @@ winzou_state_machine:
                     do: ["@sylius.order_processing.order_processor", "process"]
                     args: ["object.getOrder()"]
 
+
```
Comment (NeverResponse) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/state_machine.yml]: Should be reverted.

File: features/inventory/viewing_product_variant_with_specify_quantity_of_items_on_hand.feature [hunk 7]
```diff
@@ -0,0 +1,43 @@
+@inventory
+Feature: Seeing product's variant with specify quantity of items on hand
+    In order to see how many variants of selected product are available on hand
+    As an Administrator
+    I want to be able to see quantity of variants available on hand for a selected product
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "T-shirt banana" priced at "€12.54"
+        And the product "T-shirt banana" has "With Potato" variant priced at "€20.54"
```
Comment (lchrusciel) [inline features/inventory/viewing_product_variant_with_specify_quantity_of_items_on_hand.feature]: "T-shirt banana" has "With Potato" o.O

Comment (liverbool) [inline features/inventory/viewing_product_variant_with_specify_quantity_of_items_on_hand.feature]: 🍭🍭🍭

File: src/Sylius/Behat/Context/Ui/Shop/CartContext.php (hunk 4 above)
Comment (tuka217) [inline src/Sylius/Behat/Context/Ui/Shop/CartContext.php]: I can, but why? We pass the name of variant in test. 
We can use transformer to be sure that the variant really exist, but is it necessary?

Comment (NeverResponse) [inline src/Sylius/Behat/Context/Ui/Shop/CartContext.php]: Ofc it's not necessary, just a thought 😄

File: features/order/managing_orders/finalizing_order_payment.feature [hunk 8]
```diff
@@ -7,6 +7,7 @@ Feature: Finalizing order payment
     Background:
         Given the store operates on a single channel in "United States"
         And the store has a product "Angel T-Shirt"
+        And there is 1 item of product "Angel T-Shirt" available in the inventory
```
Comment (pjedrzejewski) [inline features/order/managing_orders/finalizing_order_payment.feature]: Why do we need this? I think inventory should be disabled by default. (tracked = false on variant)

File: src/Sylius/Behat/Context/Setup/OrderContext.php [hunk 9]
```diff
@@ -280,11 +280,16 @@ public function theCustomerBoughtProductAndProduct(ProductInterface $product, Pr
...
     /**
-     * @Given /^the customer bought (\d+) ("[^"]+" products)/
+     * @Given /^the customer bought (\d+) ("[^"]+" products)$/
+     * @Given the customer bought :quantity :product products in variant :variant
      */
-    public function theCustomerBoughtSeveralProducts($quantity, ProductInterface $product)
+    public function theCustomerBoughtSeveralProducts($quantity, ProductInterface $product, ProductVariantInterface $variant = null)
     {
-        $this->addProductVariantToOrder($this->variantResolver->getVariant($product), $product->getPrice(), $quantity);
+        if(null === $variant) {
```
Comment (pjedrzejewski) [inline src/Sylius/Behat/Context/Setup/OrderContext.php]: CS!

File: src/Sylius/Component/Core/Inventory/Updater/OnHandQuantityUpdater.php [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * @author Anna Walasek <anna.walasek@lakion.com>
+ */
+final class OnHandQuantityUpdater implements OnHandQuantityUpdaterInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function decrease($orderItems)
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Core/Inventory/Updater/OnHandQuantityUpdater.php]: I think this service should take the entire order as argument.

File: src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php [hunk 11]
```diff
@@ -146,7 +146,8 @@ public function theProductVariantShouldAppearInTheShop($productVariantCode, Prod
     }
 
     /**
-     * @When /^I want to view all variants of (this product)$/
+     * @When /^I (?:|want to )view all variants of (this product)$/
+     * @When /^I view all variants of the (product "([^"]+)")$/
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingProductVariantsContext.php]: I'm pretty sure an inside brackets should be removed. -> `(product "[^"]+")`

File: src/Sylius/Behat/Resources/config/suites/ui/inventory/inventory.yml [hunk 12]
```diff
@@ -7,27 +7,34 @@ default:
...
                 - sylius.behat.context.transform.lexical
+                - sylius.behat.context.transform.order
                 - sylius.behat.context.transform.payment
                 - sylius.behat.context.transform.product
                 - sylius.behat.context.transform.product_option
+                - sylius.behat.context.transform.product_variant
                 - sylius.behat.context.transform.shared_storage
+                - sylius.behat.context.transform.shipping_method
 
+                - sylius.behat.context.ui.admin.managing_product_variants
```
Comment (lchrusciel) [inline src/Sylius/Behat/Resources/config/suites/ui/inventory/inventory.yml]: So small suites.... Cannot we divide it into smaller pieces?

File: features/inventory/viewing_product_variant_with_specify_quantity_of_items_on_hand.feature [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "T-shirt banana" priced at "€12.54"
+        And the product "T-shirt banana" has "Yellow" variant priced at "€20.54"
+        And the product "T-shirt banana" has "Green" variant priced at "€5.54"
+        And there are 5 items of product "T-shirt banana" in variant "Yellow" available in the inventory
+        And there are 5 items of product "T-shirt banana" in variant "Green" available in the inventory
+        And the store has a product "Skirt Sheep" priced at "€34.54"
+        And the product "Skirt Sheep" has "Red" variant priced at "€5.54"
```
Comment (michalmarcinkowski) [inline features/inventory/viewing_product_variant_with_specify_quantity_of_items_on_hand.feature]: Product with a single variant is a simple product. When talking about product with a single variant we only talk about product and don't mention the variant at all. `the store has a product "X"`, `there are 5 units of product "X"  available in the inventory` etc.

File: features/inventory/viewing_product_variant_with_specify_quantity_of_items_on_hand.feature [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        And there are 5 items of product "Skirt Sheep" in variant "Red" available in the inventory
+        And the store ships everywhere for free
+        And the store allows paying with "Cash on Delivery"
+        And I am logged in as an administrator
+
+    @ui
+    Scenario: Seeing decreased quantity of product's items in selected variant after order payment
+        Given there is a customer "lucy@teamlucifer.com" that placed an order "#00000666"
+        And the customer bought 3 "Skirt Sheep" products in variant "Red"
```
Comment (michalmarcinkowski) [inline features/inventory/viewing_product_variant_with_specify_quantity_of_items_on_hand.feature]: I think we can skip creation of a second product in the background and in this scenario simply buy e.g. green banana t-shirt.

File: features/inventory/viewing_product_variant_with_specify_quantity_of_items_on_hand.feature [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+        And the customer "Lucifer Morningstar" addressed it to "Seaside Fwy", "90802" "Los Angeles" in the "United States" with identical billing address
+        And the customer chose "Free" shipping method with "Cash on Delivery" payment
+        And this order is already paid
+        When I view all variants of the product "Skirt Sheep"
+        Then I should see 2 variant in the list
```
Comment (michalmarcinkowski) [inline features/inventory/viewing_product_variant_with_specify_quantity_of_items_on_hand.feature]: It's not important for us how many variants we see there, this step should be removed.

File: features/inventory/viewing_product_variant_with_specify_quantity_of_items_on_hand.feature [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+    @ui
+    Scenario: Seeing decreased quantity of product's items from different variants after order payment
+        Given there is a customer "lucy@teamlucifer.com" that placed an order "#00000666"
+        And the customer bought 3 "T-shirt banana" products in variant "Yellow"
+        And the customer bought 2 "T-shirt banana" products in variant "Green"
+        And the customer "Lucifer Morningstar" addressed it to "Seaside Fwy", "90802" "Los Angeles" in the "United States" with identical billing address
+        And the customer chose "Free" shipping method with "Cash on Delivery" payment
+        And this order is already paid
+        When I view all variants of the product "T-shirt banana"
+        Then I should see 3 variant in the list
```
Comment (michalmarcinkowski) [inline features/inventory/viewing_product_variant_with_specify_quantity_of_items_on_hand.feature]: Same as above

File: features/inventory/viewing_product_variant_with_specify_quantity_of_items_on_hand.feature [hunk 17]
```diff
@@ -0,0 +1,43 @@
+@inventory
+Feature: Seeing product's variant with specify quantity of items on hand
+    In order to see how many variants of selected product are available on hand
+    As an Administrator
+    I want to be able to see quantity of variants available on hand for a selected product
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "T-shirt banana" priced at "€12.54"
```
Comment (michalmarcinkowski) [inline features/inventory/viewing_product_variant_with_specify_quantity_of_items_on_hand.feature]: `And the store has a "T-shirt banana" configurable product`

**Pull request closed without merging**